### PR TITLE
Preemption admission check controller. KEP update.

### LINKED ISSUE #1677: Ability to defer preemption when using admission checks
**What would you like to be added**:

The ability for an admission check to influence when preemptions happen (anytime vs after the check has passed).

**Why is this needed**:

It might be useful if an admission check could take a long time to be satisfied, so that preempting early is not very useful.
However, it's unclear whether the usefulness of this feature outweights the complexity of moving the preemption logic outside of the admission loop. Especially as we start thinking of #62 and #998. 

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind documentation

#### What this PR does / why we need it:

Preemption admission check controller. KEP update.

#### Which issue(s) this PR fixes:

Fixes #1677

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: keps/993-two-phase-admission/README.md [hunk 1]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+the scheduler will not issue the eviction during it's process instead it will set a `Pending` admission check
+that is manged by a new built-in admission check controller.
+
+The **Admission Check Controller** will:
+
+- Watch for a change in state of the workloads pending preemption.
+- Watch for workloads that are finishing execution and therefore releasing quota.
+- Watch for AdmissionCheck changes, since the preemption policy can change.
+
+Since evaluating the current preemption state and the needs of a workload requires checking the sate of all resource pool the events are rate limited the kueue's cache is used.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:269]:
```suggestion
Since evaluating the current preemption state and the needs of a workload requires checking the state of all resource pool, the events are rate limited and the kueue's cache is used.
```
Not sure if I'm interpreting the sentence correctly.
which events is this referring to?

File: keps/993-two-phase-admission/README.md [hunk 2]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+
+At every run the controller will get the list of workloads pending preemption.
+Since for some of these workloads is not necessary to issue eviction at that given point (eg. Having a pending check that uses AfterCheckPassedOrOnDemand policy) their quota reservation will be ignored.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:272]: so we will get a snapshot that doesn't contain the workloads with pending preemptions, right?

File: keps/993-two-phase-admission/README.md [hunk 3]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+Fore every other preemption pending workloads, it will check if it can fit without evicting other workloads, case in which the preemption admission check condition will be set to `Ready`.
+If eviction of other workload is still needed, an updated list candidates is created and eviction is issued for all of them.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:274]: Can you confirm if this is the implementation? and update the KEP accordingly.
In the first pass we would use the previously calculated set of preemption candidates. If this is not enough, we would recalculate the preemption candidates from scratch.

Nit: use a number list to show the steps of the algorithm.

File: keps/993-two-phase-admission/README.md [hunk 4]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+If the updated list of candidates is empty, meaning that the preemption can no longer succeed, the preemption admission check is set as `Retry`, the workload will louse it's quota reservation and requeued.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:275]:
```suggestion
If the updated list of candidates is empty, meaning that the preemption can no longer succeed, the preemption admission check is set as `Retry`, the workload will lose it's quota reservation and requeued.
```

This might be a good reason for have preemption as an AdmissionCheck, if having a permanent record for when preemptions were issued, or were not possible.

But is such permanent record useful? Given the continuous nature of preemptions, events might be a better fit. We would like to say:
1. preemptions were issued
2. preemptions were not possible.

(1) is completely continuous, so not a good fit for conditions or admission check, because lastTransitionTime wouldn't change if we issue more and more preemptions.
(2) is more permanent: once issued, we would like to requeue. But this could be represented by setting the condition `QuotaReserved` to false with reason `UnsatisfiablePreemptions`.

File: keps/993-two-phase-admission/README.md [hunk 5]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
 * After approving the workload, keep an eye on the check if it starts failing,
   fail the check and cause the workload to move back to the suspended state.
 
+### Preemption Admission Check Controller
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:257]: I have an alternative solution (please consider it, even though we might still go with your proposed solution):

Once a workload is allowed to preempt, we put it back in the scheduling queue. When poped from the queue, this workload uses a different snapshot: one that doesn't have the workload with quota reserved but not admitted.
Then we run the same algorithm we have, including preemption.

- If the checks all passed, then we can admit the workload.
- Otherwise, it goes back into the queue, probably into the "inadmissible" staging area and it can get poped back once all checks have passed.

This has the advantage that a single component is "admitting" workloads and issuing preemptions, reducing the risk for race conditions. In your proposed approach, it is possible that the preemption controller determines that a workload is admissible, at the same time that the scheduler is admitting another workload that doesn't require preemptions or checks. And the workloads that require preemption are not prioritized by the same rules as the ones that go through the scheduling queue.

**Review by alculquicondor — COMMENTED** [review]
cc @tenzen-y @mwielgus

File: keps/993-two-phase-admission/README.md [hunk 6]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+
+Because now, the time when the eviction of the preemption candidate should vary based on the preemptor state
+the scheduler will not issue the eviction during it's process instead it will set a `Pending` admission check
```
[the remark this answers is not in the document]
Comment (trasc) [reply keps/993-two-phase-admission/README.md:260]: A single condition is not sufficient, we have  "no preemption", "preemption in progress" , "preemption should start later" ... . Also the two step admission is set to "admit" when "all checks are ready" and "it has reservation" adding another "and" will make it hard to follow.  

Regarding signals ,  With or without this the eviction process take a number of API signals, (issue eviction, wait for the evicted workload to stop)

File: keps/993-two-phase-admission/README.md (hunk 2 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:272]: The workloads that don't require to issue evictions at that particular point in time are removed from the snapshot.

File: keps/993-two-phase-admission/README.md (hunk 4 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:275]: I believe that events are a better for this.

File: keps/993-two-phase-admission/README.md (hunk 3 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:274]: We don't keep track of previous lists, we are always "generating" an new list of candidates (some workload that is not present in a  previous eviction candidates list  might have ended its execution making the current list of candidates shorter).

File: keps/993-two-phase-admission/README.md (hunk 1 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:269]: ah, got it now, let me propose another rephrasing:
```suggestion
The preemption controller uses the kueue cache, since it needs to check the state of workloads admitted to the ClusterQueues.
```

Not sure if I'm leaving something out that could be relevant to the design (not sure how the events come into play).

File: keps/993-two-phase-admission/README.md (hunk 2 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:272]: Oh I see. So if there is a workload that requires preemptions and it's ready for preemption, it itself could be preempted (it's QuotaReserved would be cleared). Am I understanding correctly?

An alternative implementation (maybe weaker) would be that we only take into consideration in the snapshot the workloads that are fully admitted (not just quota reserved).

File: keps/993-two-phase-admission/README.md [hunk 7]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+
+- Watch for a change in state of the workloads pending preemption.
+- Watch for workloads that are finishing execution and therefore releasing quota.
+- Watch for AdmissionCheck changes, since the preemption policy can change.
+
+Since evaluating the current preemption state and the needs of a workload requires checking the sate of all resource pool the events are rate limited the kueue's cache is used.
+
+At every run the controller will get the list of workloads pending preemption.
+Since for some of these workloads is not necessary to issue eviction at that given point (eg. Having a pending check that uses AfterCheckPassedOrOnDemand policy) their quota reservation will be ignored.
+Fore every other preemption pending workloads, it will check if it can fit without evicting other workloads, case in which the preemption admission check condition will be set to `Ready`.
```
[the remark this answers is not in the document]
Comment (trasc) [reply keps/993-two-phase-admission/README.md:273]: The quota reserved by a workload is considered  in use by the scheduler, if another workload gets is reservation after the fist one gets to "preemption ready" , that second workload will need to trigger it's own set of evictions and so on.

File: keps/993-two-phase-admission/README.md (hunk 5 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:257]: > Once a workload is allowed to preempt, we put it back in the scheduling queue. 
The problem is that there is no guarantee that the same assignment will be available.

> This has the advantage that a single component is "admitting" workloads and issuing preemption, reducing the risk for race conditions. 
The problem is that the triggers the scheduler should respond to and the type of tasks it needs to do will dramatically increase. (I did think of having the scheduler more involved when I first started working on two-step admission but keeping the scheduler involved in only managing quota reservation is the best option in my opinion).

>  In your proposed approach, it is possible that the preemption controller determines that a workload is admissible, at the same time that the scheduler is admitting another workload that doesn't require preemptions or checks.
Actually the quota reserved by all the preemption pending  workloads is "in use" from the schedulers point of view.

>  And the workloads that require preemption are not prioritized by the same rules as the ones that go through the scheduling queue.
Indeed when it comes to the time when they are issuing eviction ,  and this one of the purpose of having preemption policies. However to get in the preemption pending state the rules are the same and during the preemption pending they are still  eligible to be evicted by a higher prio workload.

File: keps/993-two-phase-admission/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (trasc) [reply keps/993-two-phase-admission/README.md]: (fore -> for done)

File: keps/993-two-phase-admission/README.md (hunk 4 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md]: (misspell done)

File: keps/993-two-phase-admission/README.md [hunk 8]
```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
...
+the scheduler will not issue the eviction during it's process instead it will set a `Pending` admission check
+that is manged by a new built-in admission check controller.
+
+The **Admission Check Controller** will:
+
+- Watch for a change in state of the workloads pending preemption.
+- Watch for workloads that are finishing execution and therefore releasing quota.
+- Watch for AdmissionCheck changes, since the preemption policy can change.
+
+The preemption controller uses the kueue cache, since it needs to check the state of workloads admitted to the ClusterQueues.
```
Comment (tenzen-y) [inline keps/993-two-phase-admission/README.md:269]: What is the difference between the admission check controller and the preemption controller? Are those the different components? Or one component?

Comment (trasc) [reply keps/993-two-phase-admission/README.md:269]: `Admission Check Controller` is a controller that manages a specific type of `AdmissionCheck`.

`preemption controller` is short for  `Preemption Admission Check Controller`.

Comment (tenzen-y) [reply keps/993-two-phase-admission/README.md:269]: I see. Let's modify this section name, "Preemption Admission Check Controller" to the proper one.

Comment (trasc) [reply keps/993-two-phase-admission/README.md:269]: The section is named "Preemption Admission Check Controller" L257 but sure we can change L263 as well.

Comment (tenzen-y) [reply keps/993-two-phase-admission/README.md:269]: Probably, "Admission Check controller" is https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/admissioncheck_controller.go, and "preemption controller" is https://github.com/kubernetes-sigs/kueue/blob/ce7e628a1f6bb1e375790edd6fc7df3f7ac68958/pkg/controller/admissionchecks/preemption/preemption.go, right?

I wanted to clarify which sections do mention which controllers. 

So `but sure we can change L263 as well.` sounds good to me.

File: keps/993-two-phase-admission/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:260]: I agree that a condition is not a good fit. But I don't think an admission check is, either.
The concept of Preemption Ready is wrong. Once we have determined that all preemptions are done, another instance of the preemption controller can also say the same for another workload, leading to over subscription.
In other words, there could only be one instance of a preemption controller and, once determined that the preemptions are done, setting the Admitted condition should be done in the same controller. Unless we also check the Preemption status? We might as well just set the Admitted condition.

File: keps/993-two-phase-admission/README.md (hunk 5 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:257]: > The problem is that there is no guarantee that the same assignment will be available.

The same problem would be faced by a preemption controller. It needs to check that the assignment is still viable. We can have a mechanism in the scheduler logic to first try the assignment that was already calculated.

> Actually the quota reserved by all the preemption pending workloads is "in use" from the schedulers point of view

True.

> However to get in the preemption pending state the rules are the same and during the preemption pending they are still eligible to be evicted by a higher prio workload.

If two workloads have their admission checks completed, there is no guarantee that both will be able to fit after issuing preemptions. Therefore, we should prioritize them. Sure, the higher priority one could still preempt the lower priority one, but the disruption could be avoided if we had prioritized preemptions.

It's almost starting to sound like we should have a a second "preemption scheduler", instead of a "preemption controller"

**Review by alculquicondor — COMMENTED** [review]
Any implications of this update to https://github.com/kubernetes-sigs/kueue/tree/main/keps/582-preempt-based-on-flavor-order ?

File: keps/993-two-phase-admission/README.md (hunk 5 above)
Comment (tenzen-y) [reply keps/993-two-phase-admission/README.md:257]: > "preemption scheduler"

@alculquicondor Does this mean we should introduce a scheduler for preemption separate from an existing scheduler (admission scheduler?)?

I'm concerned that two schedulers can work in asynchronous cooperation as we expected.

I was thinking about the separate scheduler when I reviewed @trasc 's implementation PR. 
However, in the PR, I said we might want to somehow implement the preemption logic to the scheduling cycle instead of introducing a new scheduler.

Because I thought that it might be hard to handle different schedulers well due to some conflicts, such as lower being preempted by higher, as you mentioned. Moreover, I'm wondering if the issue would be more complex when the borrowing semantics would be related to the scheduling logic.

Honestly, I don't have a clear answer. However, I'm wondering if we should discuss the second scheduler for the preemption focused on the existing preemption logic, not including two-phase admission once away from two-phase admission with preemption.

Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:257]: I'm already thinking of leaving this change out of release-0.5, as it's highly complex and we don't have clear answers yet.

That said, I think the idea of two schedulers isn't that bad, based on:

> Actually the quota reserved by all the preemption pending workloads is "in use" from the schedulers point of view

So they might not conflict with each other. Still, it might be valuable to revisit the idea that a single scheduler has the two modes: preempting and non-preempting.

Let's give it a thought to all these ideas and come back to it in a couple of weeks.

Comment (tenzen-y) [reply keps/993-two-phase-admission/README.md:257]: > That said, I think the idea of two schedulers isn't that bad, based on:

I don't have a strong objection regarding two schedulers. I just wanted to say that I'm wondering if we shouldn't move this forward in a hurry.

> I'm already thinking of leaving this change out of release-0.5, as it's highly complex and we don't have clear answers yet.

I have the completely same opinion.

Comment (tenzen-y) [reply keps/993-two-phase-admission/README.md:257]: This is just an idea, as a byproduct, If we introduce a preemption scheduler separate from an existing scheduler, it might make providing [pluggable queue policies](https://github.com/kubernetes-sigs/kueue/issues/10) much easier since we might separate preemption logic from the flavorassigner.

**Comment by trasc** [comment]
> Any implications of this update to https://github.com/kubernetes-sigs/kueue/tree/main/keps/582-preempt-based-on-flavor-order ?

No, the changes done by that only have effect before quota reservation.

File: keps/993-two-phase-admission/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (trasc) [reply keps/993-two-phase-admission/README.md:260]: > Once we have determined that all preemptions are done, another instance of the preemption controller can also say the same for another workload, leading to over subscription.

I don't follow.  

> In other words, there could only be one instance of a preemption controller and, once determined that the preemptions are done

There is only one.

> setting the Admitted condition should be done in the same controller

It's done in the workload controller , it need to check all the Admission Checks.

File: keps/993-two-phase-admission/README.md (hunk 5 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:257]: > If two workloads have their admission checks completed, there is no guarantee that both will be able to fit after issuing preemptions. Therefore, we should prioritize them.

The priority is taken into account when creating the updated list of eviction candidates. 

Say you have a queue with capacity 5cpu and :
wl0 (prio 0,  3cpu, admitted )
wl1 (prio 1, 3cpu, quota reserved, ready for preemption)
wl2 (prio 2, 3cpu, quota reserved, ready for preemption)
During the preemption check the preemption check controller will try to:

- Get the list of candidates for wl1 while the queue has wl0 and wl2 using quota, since wl2 has higher priority the list will be empty, therefore the preemption failed for wl1. (check state `false`)
- Get the list of candidates for wl2 while the queue has wl0 and wl1 using quota, the list will contain (wl1 and wl2) they will be marked for eviction.
 
(the order in which the steps above are done is irrelevant)

I a future run (after the eviction are done) wl2 will fit. (check state `true`)

> I'm already thinking of leaving this change out of release-0.5, as it's highly complex and we don't have clear answers yet.

We have a working implemetation.

File: keps/993-two-phase-admission/README.md (hunk 4 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md]: sounds good, then I'll consider that a "record" alone is not a sufficient to decide to implement preemptions as an AdmissionCheck.

File: keps/993-two-phase-admission/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md]: > The quota reserved by a workload is considered in use by the scheduler, if another workload gets is reservation after the fist one gets to "preemption ready" , that second workload will need to trigger it's own set of evictions and so on.

Yes, that I understand.

Let me state my thinking again. There are scenarios where preemption needs to run before other admission checks have completed, because other AdmissionChecks can request preemptions.
In that case, we will issue preemptions, but even when completed, we might still need preemptions later on, when all the AdmissionChecks actually complete.

That's why I say that there might not be a concept of "preemptions are ready".

File: keps/993-two-phase-admission/README.md (hunk 5 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:257]: > (the order in which the steps above are done is irrelevant)

After reviewing your argument, I agree with this assessment.

> We have a working implemetation.

The problem is that we are touching currently stable functionality in ways that are hard to predict the outcomes. We might need to evaluate where our coverage is weak, so we can add it before changing the logic. We can't merge such a big PR lightly.

But now we are at a good point in the release cycle in which we can proceed with such an overtake. I still need a little more time to review.

**Comment by alculquicondor** [comment]
I think one important question is whether the preemption controller needs to be an admission check.

In some scenarios, the preemption check is running only once all the other admission checks have passed. This has the implication that now no other admission check can do the same, otherwise they will block each other.

So that's an argument against it being an admission check.

I'll give another pass to the proposal tomorrow and think about it more.

File: keps/993-two-phase-admission/README.md [hunk 9]
```diff
@@ -253,6 +254,40 @@ The controller implementing a particular check should:
 * After approving the workload, keep an eye on the check if it starts failing,
   fail the check and cause the workload to move back to the suspended state.
 
+### Preemption Admission Check Controller
+
+In this proposal, the time to evict the preemption candidates varies based on the preemptor state
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:259]:
```suggestion
In this proposal, the time to evict the preemption candidates varies based on the preemptor state.
```

File: keps/993-two-phase-admission/README.md [hunk 10]
```diff
@@ -253,6 +254,40 @@ The controller implementing a particular check should:
...
+- Watch for a change in state of the workloads pending preemption.
+- Watch for workloads that are finishing execution and therefore releasing quota.
+- Watch for AdmissionCheck changes, since the preemption policy can change.
+
+The preemption controller uses the kueue cache, since it needs to check the state of workloads admitted to the ClusterQueues.
+
+At every run the controller will get the list of workloads pending preemption.
+
+The workloads pending preemption are divided into:
+- `preemtingLetaer` - Workloads having at least one check that uses AfterCheckPassedOrOnDemand policy with the state `pending`.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:274]:
```suggestion
- `preemtingLater` - Workloads having at least one check that uses AfterCheckPassedOrOnDemand policy with the state `pending`.
```

File: keps/993-two-phase-admission/README.md [hunk 11]
```diff
@@ -253,6 +254,40 @@ The controller implementing a particular check should:
...
+1. Remove all workloads from the snapshot.
+2. For every workload in `preemtingNow` , in the order of their priority:
+  - If it can fit without the need of evicting other workloads
+    - Set its admission check to `Ready`
+    - Add it to the snapshot
+  - If it cannot fit
+    - Get an updated list of eviction candidates
+      - If the updated list is not empty.
+        - Issue the eviction to the candidates.
+        - Add it to the snapshot
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:287]: At this point? Shouldn't we wait for the next reconcile, given that jobs can take time to terminate?

File: keps/993-two-phase-admission/README.md [hunk 12]
```diff
@@ -253,6 +254,40 @@ The controller implementing a particular check should:
...
+      - If the updated list is empty, meaning the preemption cannot be done.
+        - Set its admission check to `Ready`
```
[the remark this answers is not in the document]
Comment (trasc) [reply keps/993-two-phase-admission/README.md]: It's "Retry", updated.

File: keps/993-two-phase-admission/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (trasc) [reply keps/993-two-phase-admission/README.md]: The preemption is finished when the workload can fit without the need of any additional evictions. Pass this point its  quota is locked and can start execution, or wait more.  The quote remains locked until this workload finishes or gets evicted.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md]: That makes me think that the existing `QuotaReserved` condition will have different semantics. And the idea of `QuotaReserved` could actually be used to signal that preemption is "ready".

But maybe that's a bit of over-engineering.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md]: Would QuotaReserved be equivalent to "preemption ready" then?

File: keps/993-two-phase-admission/README.md [hunk 13]
```diff
@@ -271,7 +271,7 @@ The preemption controller uses the kueue cache, since it needs to check the stat
 At every run the controller will get the list of workloads pending preemption.
 
 The workloads pending preemption are divided into:
-- `preemtingLetaer` - Workloads having at least one check that uses AfterCheckPassedOrOnDemand policy with the state `pending`.
+- `preemtingLeter` - Workloads having at least one check that uses AfterCheckPassedOrOnDemand policy with the state `pending`.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:274]: Later?

File: keps/993-two-phase-admission/README.md (hunk 11 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:287]: Just to be clear, you mean the snapshot, not the cache, correct?

File: keps/993-two-phase-admission/README.md [hunk 14]
```diff
@@ -253,6 +254,42 @@ The controller implementing a particular check should:
...
+    - Add it to the snapshot
+  - If it cannot fit
+    - Get an updated list of eviction candidates
+      - If the updated list is not empty.
+        - Issue the eviction to the candidates.
+        - Add it to the snapshot
+      - If the updated list is empty, meaning the preemption cannot be done.
+        - Set its admission check to `Retry`, the quota reservation will be lost and the workload placed in the queue waiting for a new QuotaReservation.
+
+**NOTE** The list of candidates is picked out from the list of workloads holding a QuotaReservation, regardless if they are fully Admitted or not.
```
Comment (alculquicondor) [inline keps/993-two-phase-admission/README.md:291]: Is QuotaReservation enough? Consider jobs waiting for ProvReq to be fulfilled could be waiting for hours. There is no point in preempting them yet.

File: keps/993-two-phase-admission/README.md (hunk 11 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:287]: Shapshot

File: keps/993-two-phase-admission/README.md (hunk 14 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md:291]: The ProvRequest Admission check can be configured with `AfterCheckPassedOrOnDemand` policy, the evictions will be done only when the provisioning is finished (and only if still needed).

In the future , if we can know that the provisioning request has no chance to succeed for long period of times, the ProvRecAC can set its state to Retry this will fully release the quota.After that time the state can be set to Pending to put wl back in the queue.

File: keps/993-two-phase-admission/README.md (hunk 13 above)
Comment (trasc) [reply keps/993-two-phase-admission/README.md]: done

File: keps/993-two-phase-admission/README.md (hunk 14 above)
Comment (alculquicondor) [reply keps/993-two-phase-admission/README.md:291]: My question is more about the candidates for eviction. Obviously, fully Admitted workloads are candidates for eviction.
Other than that, I think it should be only the ones that have checks ready or are requesting preemption themselves.
